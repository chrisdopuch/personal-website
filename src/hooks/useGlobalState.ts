import { useContext } from 'react';
import { stateCtx } from '../components/contexts/state';
import { State } from '../store';

const useGlobalState = <K extends keyof State>(key: K) => {
  const state = useContext(stateCtx);
  return state[key];
};

export default useGlobalState;
