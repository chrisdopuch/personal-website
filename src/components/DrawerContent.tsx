import { Switch } from '@material-ui/core';
import Divider from '@material-ui/core/Divider';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { SvgIconProps } from '@material-ui/core/SvgIcon';
import React, { ComponentType, SFC } from 'react';
import { Link } from 'react-router-dom';
import useDispatch from '../hooks/useDispatch';
import useGlobalState from '../hooks/useGlobalState';
import { isDarkModeKey } from '../store';

const stylesDeclarations = (theme: Theme) => {
  return createStyles({
    darkMode: {
      paddingLeft: 0,
    },
    listItemIcon: {
      marginRight: 6,
    },
    social: {
      margin: '20px 10px',
      textAlign: 'left',
    },
    toolbar: theme.mixins.toolbar,
  });
};

interface IDrawerContentProps extends WithStyles<typeof stylesDeclarations> {
  items: Array<{
    to: string;
    label: string;
    Icon: ComponentType<SvgIconProps>;
  }>;
}

export const DrawerContent: SFC<IDrawerContentProps> = (props) => {
  const { classes, items } = props;
  const isDarkMode = useGlobalState('isDarkMode');
  const dispatch = useDispatch();

  return (
    <div>
      <div className={classes.toolbar} />
      <Divider />
      <List>
        {items.map(({ label, to, Icon }) => {
          const linkProps = { to } as any;
          return (
            <ListItem
              onClick={() => dispatch({ type: 'setNavDrawerOpen', payload: false })}
              button={true}
              key={label}
              component={Link}
              {...linkProps}
            >
              <ListItemIcon>
                <Icon className={classes.listItemIcon} />
              </ListItemIcon>
              <ListItemText primary={label} />
            </ListItem>
          );
        })}
      </List>
      <Divider />
      <ListItem className={classes.darkMode}>
        <Switch
          checked={isDarkMode}
          onChange={(e) => {
            const payload = e.target.checked;
            localStorage.setItem(isDarkModeKey, payload.toString());
            return dispatch({ type: 'setDarkMode', payload });
          }}
          color="primary"
        />
        <ListItemText primary="Dark Mode" />
      </ListItem>
      <Divider />
      <div className={classes.social}>
        <a className="github-button" href="https://github.com/chrisdopuch" aria-label="Follow @chrisdopuch on GitHub">
          Follow @chrisdopuch
        </a>
      </div>
    </div>
  );
};

export default withStyles(stylesDeclarations, { withTheme: true })(DrawerContent);
