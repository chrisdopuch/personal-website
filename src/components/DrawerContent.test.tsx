import { Switch, Theme } from '@material-ui/core';
import { shallow } from 'enzyme';
import React from 'react';
import useGlobalState from '../hooks/useGlobalState';
import { DrawerContent } from './DrawerContent';

const mockDispatch = jest.fn();
jest.mock('../hooks/useDispatch', () => {
  return jest.fn(() => mockDispatch);
});

jest.mock('../hooks/useGlobalState', () => {
  return jest.fn(() => false);
});

function getDefaultProps() {
  return {
    classes: {
      darkMode: 'darkMode',
      listItemIcon: 'listItemIcon',
      social: 'social',
      toolbar: 'toolbar',
      twitter: 'twitter',
    },
    items: [],
    theme: {} as Theme,
  };
}

describe('DrawerContent', () => {
  it('renders default props', () => {
    const wrapper = shallow(<DrawerContent {...getDefaultProps()} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with `isDarkMode` true', () => {
    (useGlobalState as jest.Mock).mockImplementationOnce(() => true);
    const wrapper = shallow(<DrawerContent {...getDefaultProps()} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with items', () => {
    const MockHomeIcon = jest.fn();
    const MockHelpIcon = jest.fn();
    const MockHomePage = jest.fn();
    const MockAboutPage = jest.fn();
    const props = {
      ...getDefaultProps(),
      items: [
        { to: '/', label: 'Home', Icon: MockHomeIcon, isRouteExact: true, page: MockHomePage },
        { to: '/about', label: 'About me', Icon: MockHelpIcon, isRouteExact: false, page: MockAboutPage },
      ],
    };
    const wrapper = shallow(<DrawerContent {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("dispatches `setDarkMode` action on Switch's `onChange`", () => {
    const checked = false;
    const wrapper = shallow(<DrawerContent {...getDefaultProps()} />);

    wrapper
      .find(Switch)
      .props()
      .onChange({ target: { checked } });

    expect(mockDispatch).toBeCalledWith({ type: 'setDarkMode', payload: checked });
  });
});
