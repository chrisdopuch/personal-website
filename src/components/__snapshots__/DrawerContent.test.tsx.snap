// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerContent snapshots renders default props 1`] = `
<div>
  <div
    className="toolbar"
  />
  <WithStyles(Divider) />
  <WithStyles(List) />
  <WithStyles(Divider) />
  <div
    className="social"
  >
    <div
      className="twitter"
    >
      <Follow
        onLoad={[Function]}
        options={Object {}}
        username="chrisdopuch"
      />
    </div>
    <a
      aria-label="Follow @chrisdopuch on GitHub"
      className="github-button"
      href="https://github.com/chrisdopuch"
    >
      Follow @chrisdopuch
    </a>
  </div>
</div>
`;

exports[`DrawerContent snapshots renders with items 1`] = `
<div>
  <div
    className="toolbar"
  />
  <WithStyles(Divider) />
  <WithStyles(List)>
    <WithStyles(ListItem)
      button={true}
      component={[Function]}
      key="Home"
      onClick={[Function]}
      to="/"
    >
      <WithStyles(ListItemIcon)>
        <mockConstructor />
      </WithStyles(ListItemIcon)>
      <WithStyles(ListItemText)
        primary="Home"
      />
    </WithStyles(ListItem)>
    <WithStyles(ListItem)
      button={true}
      component={[Function]}
      key="About me"
      onClick={[Function]}
      to="/about"
    >
      <WithStyles(ListItemIcon)>
        <mockConstructor />
      </WithStyles(ListItemIcon)>
      <WithStyles(ListItemText)
        primary="About me"
      />
    </WithStyles(ListItem)>
  </WithStyles(List)>
  <WithStyles(Divider) />
  <div
    className="social"
  >
    <div
      className="twitter"
    >
      <Follow
        onLoad={[Function]}
        options={Object {}}
        username="chrisdopuch"
      />
    </div>
    <a
      aria-label="Follow @chrisdopuch on GitHub"
      className="github-button"
      href="https://github.com/chrisdopuch"
    >
      Follow @chrisdopuch
    </a>
  </div>
</div>
`;
